src/abstractdataframe/selection.jl

```diff
@@ -0,0 +1,475 @@
+    end
+    transformed_cols[newname] = nothing
+end
+
+"""
+    select!(df::DataFrame, inds...)
+
+Mutate `df` in place to retain only columns specified by `inds...` and return it.
+
+Arguments passed as `inds...` can be any index that is allowed for column indexing.
+In particular, regular expressions, `All`, `Between`, and `Not` selectors are supported.
+
+Columns can be renamed using the `old_column => new_column_name` syntax,
+and transformed using the `old_column => fun => new_column_name` syntax.
+`new_column_name` must be a `Symbol`, and `fun` a function or a type.
+If `old_column` is a `Symbol` or an integer then `fun` is applied to each element
+(row) of `old_column`.
+Otherwise `old_column` can be any column indexing syntax, but in this case `fun`
+will be passed a `NamedTuple` representing each row, holding only
+the columns specified by `old_column`.
+
+Column transformation can also be specified using the short `old_column => fun` form.
+In this case, `new_column_name` is automatically generated as `\$(old_column)_\$(fun)`.
+Up to three column names are used for multiple input columns and they are joined
+using `_`; if more than three columns are passed then the name consists of the
+first two names and `etc` suffix then, e.g. `[:a,:b,:c,:d] => fun` produces
+the new column name `a_b_etc_fun`.
+
+If a collection of column names is passed to `select!` then requesting duplicate column
+names in target data frame are accepted (e.g. `select!(df, [:a], :, r"a")` is allowed)
+and only the first occurrence is used. In particular a syntax to move column `:col`
+to the first position in the data frame is `select!(df, :col, :)`.
+On the contrary, output column names of renaming, transformation and single column
+selection operations must be unique, so e.g. `select!(df, :a, :a => :a)` or
+`select!(df, :a, :a => sin => :a)` are not allowed.
+
+Note that including the same column several times in the data frame via renaming
+when `copycols=false` will create column aliases. An example of such a situation is
+`select!(df, :a, :a => :b, :a => :c, copycols=false)`.
+
+# Examples
+```jldoctest
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select!(df, 2)
+3×1 DataFrame
+│ Row │ b     │
+│     │ Int64 │
+├─────┼───────┤
+│ 1   │ 4     │
+│ 2   │ 5     │
+│ 3   │ 6     │
+
+julia> df = DataFrame(a=1:3, b=4:6)
+3×2 DataFrame
+│ Row │ a     │ b     │
+│     │ Int64 │ Int64 │
+├─────┼───────┼───────┤
+│ 1   │ 1     │ 4     │
+│ 2   │ 2     │ 5     │
+│ 3   │ 3     │ 6     │
+
+julia> select!(df, :a => Row(sin) => :c, :b)
+3×2 DataFrame
+│ Row │ c        │ b     │
+│     │ Float64  │ Int64 │
+├─────┼──────────┼───────┤
+│ 1   │ 0.841471 │ 4     │
+│ 2   │ 0.909297 │ 5     │
+│ 3   │ 0.14112  │ 6     │
+
+julia> select(df, :, [:c, :b] => x -> x.c + x.b .- sum(x.b) / length(x.b))
```

Comment:
Shouldn't this show off select`!`?

Reply:
It should :). Thank you for spotting this (in general: I am holding off with a full docstring and test updates before I get a sign-off from @nalimilan that we go for the proposed design).